### DISCUSSION TITLE
[WebRTC] closed: deadline exceeded while waiting connection

### DISCUSSION THREAD

**Discussion by derekcct2** [bluenviron/mediamtx#3220] [Questions] [answered]
### Question

Before version 1.3.0 of mediaMTX, 
if the webrtcICEHostNAT1to1IPs was set to a fixed IP, such as 192.168.1.100, everything worked fine. 
However, starting from version 1.3.0, webrtcICEHostNAT1to1IPs was changed to webrtcAdditionalHosts. 
If webrtcAdditionalHosts is set to 192.168.1.100, an error occurs stating "[WebRTC] closed: deadline exceeded while waiting connection," and the video cannot be seen through WebRTC. 
Is this behavior normal?
My goal is to set webrtcAdditionalHosts to a fixed IP, DNS, and internal IP, but whenever I include an IP in the settings, this error occurs. Did I configure something incorrectly?

**Comment by derekcct2** [comment]
Here are the configurations and logs from when I ran the executable file of MediaMTX version 1.6.0 on Windows.
In the mediamtx.yml, whenever I change webrtcAdditionalHosts: [192.168.70.46] to webrtcAdditionalHosts: [], I can successfully view the stream's video from http ://192.168.70.46:8889/test.
[log.zip](https://github.com/bluenviron/mediamtx/files/14930991/log.zip)

**Comment by aler9** [comment, marked as answer]
Hello, the other change that was introduced in v1.3.0 was that WebRTC peer connections are now established by using a single, fixed port, that can be set through `webrtcLocalUDPAddress` or `webrtcLocalTCPAddress` and by default is `8189`.

This port must be opened or forwarded to users in the same way that port in `webrtcAddress` is (`8889`), otherwise a connection cannot be established.

The advantage of this method lies in the fact that a STUN server is not necessary anymore and therefore no third-party service is contacted whenever you want to use WebRTC.

If you liked the old method better, in which a STUN server and a random UDP port were used in order to establish connection, and no port forwarding was needed, you can re-enable it by following instructions in the README:

```yml
# STUN servers allows to obtain and share the public IP of the server.
webrtcICEServers2:
  - url: stun:stun.l.google.com:19302
```

random port and public IP address will appear in the SDP:

```
a=candidate:3901994122 1 udp 1694498815 x.x.x.95 48537 typ srflx raddr 0.0.0.0 rport 48537
a=candidate:3901994122 2 udp 1694498815 x.x.x.95 48537 typ srflx raddr 0.0.0.0 rport 48537
```

**Reply by derekcct2** [reply]
Thank you for your explanation. 
As you mentioned, when I set webrtcICEServers2 to stun:stun.l.google.com:19302, I am able to successfully establish a WebRTC connection. However, I am curious about something. When webrtcAdditionalHosts is not set, the WebRTC connection works fine regardless of whether webrtcICEServers2 is set or not. 
But if webrtcAdditionalHosts is set to a fixed IP other than 127.0.0.1, then webrtcICEServers2 must be set, otherwise, the connection fails. Why is this happening? 
I am running WebRTC on Windows as an executable file, so there should be no firewall blocking the 8189 port.
Does this mean that if I want to set webrtcAdditionalHosts, whether it's a fixed IP or a DNS, I must also fill in webrtcICEServers2?

**Reply by aler9** [reply]
> Does this mean that if I want to set webrtcAdditionalHosts, whether it's a fixed IP or a DNS, I must also fill in webrtcICEServers2?

No, it means that your users are unable to access the server through host contained in `webrtcAdditionalHosts` and the port in `webrtcLocalUDPAddress`, which is a UDP port and might be as well blocked by the Windows firewall by default. I cannot say anything else since a network dump would be needed for further analysis.

When you fill `webrtcICEServers2`, you are enabling the legacy connection mechanism, which runs in parallel to the new one. Whichever is the first to work, it is used for exchanging packets.

[a comment by an automated account is not shown]
